firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java

```diff
@@ -296,27 +338,61 @@ AppDistributionReleaseInternal getCachedLatestRelease() {
         AlertDialog.BUTTON_POSITIVE,
         context.getString(R.string.update_yes_button),
         (dialogInterface, i) ->
-            // show download progress in notification manager
+        // show download progress in notification manager
+        {
+          synchronized (updateTaskLock) {
             updateApp(true)
-                .addOnSuccessListener(unused -> updateAlertDialogTask.setResult(null))
-                .addOnFailureListener(updateAlertDialogTask::setException));
+                .addOnProgressListener(
+                    progress -> cachedUpdateToLatestReleaseTask.updateProgress(progress))
+                .addOnSuccessListener(unused -> cachedUpdateToLatestReleaseTask.setResult())
+                .addOnFailureListener(cachedUpdateToLatestReleaseTask::setException);
+          }
+        });
 
     alertDialog.setButton(
         AlertDialog.BUTTON_NEGATIVE,
         context.getString(R.string.update_no_button),
         (dialogInterface, i) -> {
           dialogInterface.dismiss();
-          updateAlertDialogTask.setException(
-              new FirebaseAppDistributionException(
-                  Constants.ErrorMessages.UPDATE_CANCELED,
-                  FirebaseAppDistributionException.Status.INSTALLATION_CANCELED));
+          synchronized (updateTaskLock) {
```

Comment:
I think it might be worth extracting the updateProgress setting and the `setException` into a method since they are being called in a couple of places. Added advantage - will reduce the number of synchronization blocks.

Reply:
The updateProgress with a setException in the same block only happens once. the other time the pieces of code are close they are in a failure listen and on progress listener. So I am not sure it makes much sense to take it out since it seems this combo only happens once. In other places there are either a setTaskCompletion or a setResult.